open_seq2seq/models/text2speech.py

```diff
@@ -0,0 +1,415 @@
+    summary = tf.Summary.Image(encoded_image_string=s.getvalue(),
+                                height=int(fig.get_figheight()*300),
+                                width=int(fig.get_figwidth()*300))
+    summary = tf.Summary.Value(tag=tag, image=summary)
+    plt.close(fig)
+
+    return summary
+  else:
+    if append:
+      name = '{}/Output_step{}_{}_{}.png'.format(logdir, train_step, number, append)
+    else:
+      name = '{}/Output_step{}_{}.png'.format(logdir, train_step, number)
+    if logdir[0] != '/':
+      name = "./"+name
+    #save
+    fig.savefig(name, dpi=300)
+
+    plt.close(fig)
+    return None
+
+def save_audio(magnitudes, logdir, step, mode="train", number=0, save_to_tensorboard=False):
+  # magnitudes = np.exp(mag_spec)
+  signal = griffin_lim(magnitudes.T**1.2)
+  file_name = '{}/sample_step{}_{}_{}.wav'.format(logdir, step, number, mode)
+  if logdir[0] != '/':
+    file_name = "./"+file_name
+  write(file_name, 22050 ,signal)
+  if save_to_tensorboard:
+    tag = "{}_audio".format(mode)
+    s = StringIO()
+    write(s, 22050, signal)
+    # summary = tf.Summary.Audio(sample_rate = 22050,
+                                # num_channels = 1,
+                                # length_frames = len(signal),
+                                # encoded_audio_string=s.getvalue(),
+                                # content_type="audio/wav")
+    summary = tf.Summary.Audio(encoded_audio_string=s.getvalue())
+    summary = tf.Summary.Value(tag=tag, audio=summary)
+    return summary
+  else:
+    file_name = '{}/sample_step{}_{}_{}.wav'.format(logdir, step, number, mode)
+    if logdir[0] != '/':
+      file_name = "./"+file_name
+    write(file_name, 22050 ,signal)
+    return None
+
+def griffin_lim(magnitudes, n_iters=50):
+  """
+  PARAMS
+  ------
+  magnitudes: spectrogram magnitudes
+  stft_fn: STFT class with transform (STFT) and inverse (ISTFT) methods
+  """
+
+  phase = np.exp(2j * np.pi * np.random.rand(*magnitudes.shape))
+  complex_spec = magnitudes * phase
+  signal = librosa.istft(complex_spec)
+
+  for i in range(n_iters):
+    _, phase = librosa.magphase(librosa.stft(signal, n_fft=1024))
+    complex_spec = magnitudes * phase
+    signal = librosa.istft(complex_spec)
+  return signal
+
+class Text2Speech(EncoderDecoderModel):
+  @staticmethod
+  def get_required_params():
+    return dict(EncoderDecoderModel.get_required_params(), **{
+      'save_to_tensorboard': bool,
+    })
+
+  def __init__(self, params, mode="train", hvd=None):
+    super(Text2Speech, self).__init__(params, mode=mode, hvd=hvd)
+    self.eval_steps = -1
+    self.train_steps = -1
+    self.save_to_tensorboard = self.params["save_to_tensorboard"]
+
+
+  def _create_decoder(self):
+    self.params['decoder_params']['num_audio_features'] = (
+      self.get_data_layer().params['num_audio_features']
+    )
+    return super(Text2Speech, self)._create_decoder()
+
+  def _build_forward_pass_graph(self, input_tensors, gpu_id=0):
```

Comment:
The idea of using EncoderDecoderModel is that you won't reimplement this function yourself. Can you list the reasons why the design of the base function is not flexible enough for your model so that we can either modify the base function or change your encoder-decoder to comply with it?

Reply:
The only change is when I can return from _build_forward_pass_graph and subsequently have access to inside maybe_print_logs, finalize_evaluation, and finalize_inference. I suppose in the final revision only returning the final sample would be ok. In the meantime, I would like to have have access to other model outputs (like as attention alignments) for debugging.

Reply:
Ok, so you will remove it later? And even if you will do, do you think it will be useful to have some API for returning more tensors from various places?

Reply:
Actually, on second thought, I might also need the final sequence length since there is no explicit EOS token for speech. So it might be helpful to allow the user to return more tensors.

Why not just return loss, decoder_output? Where decoder_output is the dictionary returned from the decoder.

Reply:
The problem is that whatever is returned as model.output_tensors will be evaluated on each session run and we don't necessary want to evaluate all of those things. But I think we can just rename the decoder["samples"] to something more meaningful and just pass whatever we want to actually appear in the maybe_print_logs and other function there. I don't think there is any restriction on it being a single tensor or if there is one it should be easily removed. Will that be convenient for you?

Reply:
This is true, but can't the evaluate function just ignore those extra parameters? They will not get passed to finalize evaluation, which seems to the bottleneck imo. I could be very wrong here.

Regardless, your suggestion is also a good idea. We can switch to that. It seems to be similar to my idea, since the model designer will decide what to return in decoder_output. But now, the model designer will just put this into the new data structure.

Reply:
Renamed samples to outputs in PR #159. Feel free to just pass any list of tensors there. But note that they will all be evaluated before passing to maybe_print_logs and evaluate functions, so you probably want to keep only necessary ones in the final version.

Reply:
Can we change 
https://github.com/NVIDIA/OpenSeq2Seq/blob/45def43e1809a446e4df2fd1d3fe4a6860c558ac/open_seq2seq/models/encoder_decoder.py#L159

from 
if self.mode == "train":
  decoder_input['target_tensors'] = target_tensors
to 
if self.mode != "infer":
  decoder_input['target_tensors'] = target_tensors

is there a specific modality that cannot accept target tensors during eval? It seems helpful to give the target sequence length at eval time so the decoder knows when to stop decoding.

Reply:
The idea is that evaluation graph should be the same as inference except for the accuracy calculation. Since during inference you don't have targets, they are not passed to the model during evaluation.